src/cli/cli_tcp.cpp

```diff
@@ -0,0 +1,522 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a TCP CLI tool.
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TCP_ENABLE
+
+#include "cli_tcp.hpp"
+
+#include <openthread/tcp.h>
+
+#include "cli/cli.hpp"
+#include "common/encoding.hpp"
+#include "common/timer.hpp"
+
+namespace ot {
+namespace Cli {
+
+constexpr TcpExample::Command TcpExample::sCommands[];
+
+TcpExample::TcpExample(Interpreter &aInterpreter)
+    : mInterpreter(aInterpreter)
+    , mEndpointInitialized(false)
+    , mListenerInitialized(false)
+    , mEndpointConnected(false)
+    , mSendBusy(false)
+    , mBenchmarkBytesTotal(0)
+    , mBenchmarkLinksLeft(0)
+{
+    memset(&mEndpoint, 0x00, sizeof(mEndpoint));
+    memset(&mListener, 0x00, sizeof(mListener));
+}
+
+otError TcpExample::ProcessHelp(Arg aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgs);
+
+    for (const Command &command : sCommands)
+    {
+        mInterpreter.OutputLine(command.mName);
+    }
+
+    return OT_ERROR_NONE;
+}
+
+otError TcpExample::ProcessInit(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+    size_t  receiveBufferSize;
+
+    VerifyOrExit(aArgs[0].IsEmpty() || aArgs[1].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
```

Comment:
We can remove this line. Will be covered below.

Can possibly add `VerifyOrExit(aArgs[1].IsEmpty(), error = OT_ERROR_INVALID_ARGS)` after the `SuccessOrExit(error = aArgs[0].ParseAsUint32(windowSize));` (to make sure we have only one agr)